### ISSUE TITLE
gatk-register command

### ISSUE THREAD

**Issue by peflanag** [ngs-fzb/MTBseq_source#29]
Hi, 

I have installed MTBseq via conda and I am using the the gatk-register command to add GATK to the conda environment but I am told the command is not found.

Any ideas?

Cheers!

**Comment by peflanag** [comment]
Update. I used gatk3-register and it worked but when I try "MTBseq --check" it says gatk isnt installed or in path. 

Any Ideas?


Cheers!

**Comment by pvanheus** [comment]
@peflanag gatk3-register puts `GenomeAnalysisTK.jar` in `$CONDA_PREFIX/opt/gatk-3.8` whereas `MTBseq` looks for it in `$CONDA_PREFIX/share/mtbseq-1.0.3-1/opt`. My workaround for this is:

```
export GATKDIR=$(dirname $(realpath $(which MTBseq)))/opt
test -e $GATKDIR || mkdir $GATKDIR
cp $CONDA_PREFIX/opt/gatk-3.8/GenomeAnalysisTK.jar $GATKDIR
```

**Comment by cutpatel** [comment]
@peflanag 
This is due to the changes of GATK 3.8 to GATK 4. The conda command used to be "gatk" for 3.8 and is now gatk3, as GATK 4 is now "gatk". Also the gatk-register changed to gatk3-register for 3.8.
MTBseq expects an executable "gatk" which is not present from the gatk3-register command, only "gatk3".

For the current version to run MTbseq you just need to rename "gatk3" to "gatk" here: miniconda3/bin/gatk3 or if you are using an seperate environment miniconda3/envs/<YOUR_ENVIRONMENT>/bin/gatk3

@pvanheus 
MTBseq is first checking for an executable "gatk" available in the PATH, which should be the case if in your activated conda environment GATK is registered, with the aforementioned limitations. If not successfull, second MTBseq is checking its own opt/ directory where the GATK .jar or a symlink to it should be present if you followed the instructions how to install from the source.

You can actually see that the "gatk3" executable in the /bin directory is actually a symlink to the /opt/gatk-3.8/ directory and the execution script gatk3.py.

We understand the need that this must be changed as all new miniconda installations will face this problem.

**Comment by peflanag** [comment]
Hi, many thanks for the info and sorry in replying. I have been on annual leave for the last 3 weeks and only seeing this now. Hopefully I will get it running this week!

**Comment by cutpatel** [comment]
FYI
We have introduced the changes to gatk3 in the version 1.0.4. It should work now directly from bioconda after using the gatk3-register command. Drop me a message if not :).

**Comment by peflanag** [comment]
Oh awesome cheers!

**Comment by Firedrops** [comment]
Just tagging on for a similar query;

If I already have gatk(4) installed (say, in `/gatk/` via `gradlew`), is it still necessary to download the full .jar? Or are there ways to use the existing installation?

**Comment by cutpatel** [comment]
Gatk4 is not working with our pipeline. I have no experience on grandlew, sorry. With bioconda you can now have both versions in parallel. gatk3 and gatk (4).

**Comment by Aanabel** [comment]
Hi, I have problems with GATK, when I do the MTBseq --check says 
<INFO>	[2019-07-05 13:36:30]	Found perl module: MCE
<INFO>	[2019-07-05 13:36:30]	Found perl module: Statistics::Basic
<INFO>	[2019-07-05 13:36:30]	Found bwa in your PATH!
<INFO>	[2019-07-05 13:36:30]	Found samtools in your PATH!
<INFO>	[2019-07-05 13:36:30]	**Found gatk3 in your PATH!**
<INFO>	[2019-07-05 13:36:30]	Found picard in your PATH!

But when I run --step TBrefine, only the .gatk.bamlog file appears with a message, GATK jar file not found. Have you run "gatk3-register" ?.
 
When I did the register:
**gatk3-register /home/anabel/anaconda3/bin/GenomeAnalysisTK** --> is this the right way?, because I have been trying several ways to do the register, but for some reason it's not working. 
ENV_PREFIX /home/anabel/anaconda3
Extension specifed is not expected: GenomeAnalysisTK
  Usage: gatk3-register /path/to/GenomeAnalysisTK[-3.8.tar.bz2|.jar]

I even uninstalled and installed Gatk and MTBseq again.

I already tried to change only the gatk3 file name as in the comment above, but in the MTBseq --check says that it is not in the path.

Thanks!

**Comment by cutpatel** [comment]
@Aanabel 

Please see the description in the manual. In short, yes you first need to use the gatk3-register command. But not on the traget you specified, this is itself only a symlink  to `GenomeAnalysisTK -> ../opt/gatk-3                .8/gatk3.py`.
You need to download the gatk version from the link in the manual, best unzip it, and run the register command on the GenomeAnalysisTK.jar file.
You do not need to rename anything anymore!

**Comment by Firedrops** [comment]
> Gatk4 is not working with our pipeline. I have no experience on grandlew, sorry. With bioconda you can now have both versions in parallel. gatk3 and gatk (4).

I will stick to gatk3 considering the massive size difference (~14mb vs ~380mb).

I'm now running into another error
`gatk3-register /miniconda/dependencies/GenomeAnalysisTK[-$PKG_VERSION.tar.bz2|.jar]`  returns `bash: .jar]: command not found`
where my gatk is at `/miniconda/dependencies/GenomeAnalysisTK-3.8-1-0-gf15c1c3ef`.

java -version returns 
```
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (build 1.8.0_212-8u212-b03-2~deb9u1-b03)
OpenJDK 64-Bit Server VM (build 25.212-b03, mixed mode)
```

and just `jar` at root `/` returns
```
Usage: jar {ctxui}[vfmn0PMe] [jar-file] [manifest-file] [entry-point] [-C dir] files ...
Options:
    -c  create new archive
    -t  list table of contents for archive
    -x  extract named (or all) files from archive
    -u  update existing archive
    -v  generate verbose output on standard output
    -f  specify archive file name
    -m  include manifest information from specified manifest file
    -n  perform Pack200 normalization after creating a new archive
    -e  specify application entry point for stand-alone application 
        bundled into an executable jar file
    -0  store only; use no ZIP compression
    -P  preserve leading '/' (absolute path) and ".." (parent directory) components from file names
    -M  do not create a manifest file for the entries
    -i  generate index information for the specified jar files
    -C  change to the specified directory and include the following file
If any file is a directory then it is processed recursively.
The manifest file name, the archive file name and the entry point name are
specified in the same order as the 'm', 'f' and 'e' flags.

Example 1: to archive two class files into an archive called classes.jar: 
       jar cvf classes.jar Foo.class Bar.class 
Example 2: use an existing manifest file 'mymanifest' and archive all the
           files in the foo/ directory into 'classes.jar': 
       jar cvfm classes.jar mymanifest -C foo/ .
```

**Comment by cutpatel** [comment]
@Firedrops 

gatk3-register needs to be run directly on either the zipped archive or the unpacked GenomeAnalysisTK.jar file. Download from link in the manual, unpack and run the gatk3-register /path/to/your/GenomeAnalysisTK.jar which will copy the GenomeAnalysisTK.jar into the miniconda environment.

**Comment by Aanabel** [comment]
Hi!
This time, I just did as the manual says, but for some reason the gatk3-register file does not appear in my anaconda3 bin, as before. How can I get it?
Then when I run it on unzipped archive GenomeAnalysisTK.jar, it says gatk3-register: command/order not found. 

I downloaded as recommended, from MANUAL link. but this time I can not find the file. What do you recommend? or What I have to do?

**Comment by cutpatel** [comment]
@aanabel
gatk3-register is part of the bioconda gatk3 and not of mtbseq.
Try to remove mtbseq from conda and all its dependencies and reinstall, so that all dependencies are also reloaded. I would also use a separate environment of conda.

**Comment by Aanabel** [comment]
I already did it, I ran this command

**gatk-register /home/anabel/GenomeAnalysisTK-3.8-0-ge9d806836/GenomeAnalysisTK.jar** 
ENV_PREFIX /home/anabel/anaconda3/envs/mycondaenv
Processing GenomeAnalysisTK.jar as *.jar
jar file specified matches expected version
Copying GenomeAnalysisTK.jar to /home/anabel/anaconda3/envs/mycondaenv/opt/gatk-3.8

BUT! MTBseq --check says  
<INFO>	[2019-07-12 16:13:17]	Found perl module: MCE
<INFO>	[2019-07-12 16:13:17]	Found perl module: Statistics::Basic
<INFO>	[2019-07-12 16:13:17]	Found bwa in your PATH!
<INFO>	[2019-07-12 16:13:17]	Found samtools in your PATH!
<ERROR>	[2019-07-12 16:13:17]	gatk3 is **not** installed or not in your PATH! 

So, I'm a little bit confused because I think the symlink has been successfully, but when doing MTBseq --check, the error keep showing up. 

Sincerely, I haven't run a test yet. anyways I'll keep going. But thanks for your help!

**Comment by cutpatel** [comment]
gatk-register or gatk3-register?

Do you find in the mycondaenv/bin/ gatk or gatk3? If you have gatk try rename it to gatk3.

There seem to be gatk version problems depending on the conda version used.

**Comment by cutpatel** [comment]
This issue seems to be resolved. If there are still problems please open another issue.
